WIP: Sitemap

b5 opened the pull request:
Should get us ready to test our first milestone 😄.

Easiest way to play with this one is via the restored `walk sitemap` command, which accepts a configuration and generates a sitemap.json file whenever you `SIGKILL` the process (control-C).

This PR introduces [badger](https://github.com/dgraph-io/badger) as a key-value store that sitemap can rely on to store requests that can be built up as Resources come in, and output the final sitemap by iterating keys with a matching prefix to generate the sitemap itself. This Approach also makes versioning different sitemap crawls by varying up the prefix a near-term possibility.

Based on previous work with sitemap building & massive amounts of RAM consumption, I feel like Badger will provide a _lot_ of mileage for this project moving forward. This PR will eventually include a Badger implementation of the `RequestStore` interface.

b5 committed:
feat(Sitemap): initial sitemap ResourceHandler & Generator

Added a first pass at a Sitemap ResourceHandler, and a "sitemap" subcommand to test it out

I've added a dependency on the Badger key/value sture, which seems like a number of other
pieces of walk could utilize. It's a somewhat opinionated choice, but definitely does
the trick.

b5 committed:
chore(ci): add badger dependency

b5 committed:
test: standardize test suite, add initial coordinator test

This gets us moving in the right direction in terms of having a maintainable set of tests
I've also added OnPush & OnPop functions to MemQueue, because it's a great place to
inject testing logic. Making sure the coorinator *never* submits duplicate URLS to the
queue is, well, nice.

closes #4

Mr0grog commented on lib/coordinator.go:
```diff
@@ -79,6 +79,15 @@ func NewCoordinator(cfg *CoordinatorConfig, q Queue, frs RequestStore, rh []Reso
 	return c
 }
 
+// SetHandlers confirgures the coordinator's resource handlers
```
Typo: confirgures → configures

Mr0grog commented on lib/lib_test.go:
```diff
@@ -0,0 +1,66 @@
+
+// TestCase is a simlation of a domain for crawling, constructed from a directory with a
+// standardized structure ()
+type TestCase struct {
```
I feel like this could do with a more specific name — not all tests will necessarily need this, but it makes it sound like they do. Maybe `CrawlableTestCase`?

Mr0grog commented on lib/coordinator_test.go:
```diff
@@ -0,0 +1,36 @@
+	"testing"
+)
+
+func TestCoordinatorRequeue(t *testing.T) {
```
Feels like this name could be descriptive (or come with a descriptive comment). If I’m understanding right, we are testing that a given URL won’t get queued more than once in the same crawl. From the name, I had initially assumed this would be about making sure a URL *did* get requeued the response was bad.

Mr0grog commented on lib/request.go:
```diff
@@ -26,6 +26,8 @@ const (
 	// RequestStatusQueued indicates this Request is queued for fetching
 	RequestStatusQueued
+	// RequestStatusRequesting indicates this Request is currently being fetched
+	RequestStatusRequesting
```
Is this for future use somewhere? I don’t see it referenced elsewhere here.

b5 replied:
Yep, planning on using this in the future, in particular based on our conversations about moving in-progress Requests to their own queue for failure handling

Mr0grog commented:
Is this ready for (or desiring) review?

Mr0grog commented on lib/sitemap/sitemap.go:
```diff
@@ -0,0 +1,208 @@
+	db     *badger.DB
+}
+
+// NewGenerator carets a generator from a given prefix & badger.DB connection
```
Typo: carets → creates

Mr0grog commented on lib/sitemap/sitemap.go:
```diff
@@ -0,0 +1,208 @@
+
+	value, err := json.Marshal(me)
+	if err != nil {
+		log.Debugf("error encoding map entry: %s", err.Error())
```
Should we be returning here?

Mr0grog commented on lib/sitemap/sitemap.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+// NormalizeURLString canonicalizes a URL
+func NormalizeURLString(rawurl string) (string, error) {
```
Feels like this should move into `lib` and `lib.NormalizeURLString()` should just be `lib.NormalizeUrl()` and then this should call that (instead of calling Purell directly).

b5 replied:
Yes. I think we need to consider url normalization as a configuration-level thing and establish how that should work. People have, like, _opinions_ on URL normalization. It'd be nice to have normalization as a lib function with clear docs (and config convention) that pieces of walk can make easy use of. I think this should also be a separate PR

Mr0grog replied:
> People have, like, *opinions* on URL normalization.

What? Preposterous!

b5 commented:
I'd love a mid-point checkin 😄. Generally I like to formally "request review" when ready for such a thing, but I'm a _huge_ fan of mid-way comments to keep PRs on the right track.

In my head at least two things are left to do in this PR:
- [x] land sitemap generation as a configuration setting (one of the main goals of the project is to do more with a single request)
- [x] properly implement the parallelization setting on the refactored worker (without it we're stuck on a single lane of requests. No bueno)

I'm thinking redirect handling will be better handled as a separate PR.

Mr0grog commented on lib/sitemap/sitemap.go:
```diff
@@ -0,0 +1,208 @@
+		return err
+	}
+
+	data, err := json.MarshalIndent(sm, "", "  ")
```
Might be a good idea to ultimately make this a command line option.

Farther afield, I think we might have talked about maybe making the maps newline-delimited JSON streams so that something reading one could parse it entry-by-entry rather than having to read the whole sitemap file in.

b5 replied:
Totally agreed. For now I'd like to keep it this way, and mark these two suggestions as "nice-to-haves" in new issues.

Mr0grog replied:
Sounds like a plan.

Mr0grog reviewed (commented):
D'oh, I just realized you marked this as a WIP after I went through and did a quick review, so I’m just posting what I did as some minor comments.

> the restored `walk sitemap` command, which accepts a configuration and generates a `sitemap.json` file whenever you `SIGKILL` the process (control-C).

Is the `SIGKILL` requirement just because we don’t have a protocol for the coordinator telling all the handlers that it’s done? (Maybe we should add that?)

> This Approach also makes versioning different sitemap crawls by varying up the prefix a near-term possibility.

👍 👍 Seems like it’s the easy way to have multiple sites being mapped in parallel, too.

> This PR will eventually include a Badger implementation of the RequestStore interface.

Also 👍 👍

Mr0grog commented:
(Should have said: mostly looks awesome! Hence my commends being pretty confined to naming and not much functionality or architecture :P)

Mr0grog commented:
> I'm thinking redirect handling will be better handled as a separate PR.

Sounds good to me.

Mr0grog commented:
>> the restored `walk sitemap` command, which accepts a configuration and generates a `sitemap.json` file whenever you `SIGKILL` the process (control-C).
>
> Is the `SIGKILL` requirement just because we don’t have a protocol for the coordinator telling all the handlers that it’s done? (Maybe we should add that?)

Oh, hmmmm, so also, it seems like if you don’t have a config that trips over https://github.com/qri-io/walk/pull/2#discussion_r209515313, then this will exit automatically without generating a sitemap. That seems problematic :\

b5 committed:
refactor(lib,sitemap): integrate PR feedback

b5 committed:
refactor(lib): parallel crawler, maor config, move sitemap into lib

a bunch of refactoring:
* moved the sitemap package into lib, just easier for now
* added sitemap as a configurable ResourceHandler
* removed sitemap command, it's now merged into 'walk start' with a config that specifies sitemap
* added new 'ResourceFinalizer' opt-in interface that gets called on completion, sitemap
  resourcehandlers use this as a chance to generate sitemap.json
* added back parallelism to localWorker
* plumbed in badger configuration

b5 commented:
Bunch of updates for ya' @Mr0grog. To test I've been running `walk start --config=config.json` with this config.json:
```json
{
    "Coordinator": {
        "Domains": [
            "https://datatogether.org"
        ],
        "IgnorePatterns": [],
        "Seeds": [
            "https://datatogether.org"
        ],
        "StopAfterEntries": 0,
        "UnfetchedScanFreqMilliseconds": 30000,
        "BackupWriteInterval": 500,
        "BackoffResponseCodes": [
            403
        ],
        "DoneScanMilli": 1000
    },
    "Queue": {
        "Type": "local"
    },
    "RequestStore": {
        "Type": "local"
    },
    "Workers": [
        {
            "Type": "local",
            "Polite": false,
            "UserAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36",
            "Parallelism": 5,
            "RecordRedirects": true,
            "RecordResponseHeaders": true,
            "DelayMilli": 100
        }
    ],
    "ResourceHandlers": [{
        "type" : "sitemap",
        "prefix" : "sm_01",
        "destPath" : "sitemap.json"
    },{
        "type" : "cbor",
        "destPath" : "requests"
    }]
}
```

Mr0grog commented:
Awesome! Dunno if I will have time to look tonight; will try and get it done sometime tomorrow.

Mr0grog commented on cmd/start.go:
```diff
@@ -36,9 +36,13 @@ var StartCmd = &cobra.Command{
+		for _, rh := range coord.ResourceHandlers() {
+			if finalizer, ok := rh.(lib.ResourceFinalizer); ok {
+				if err := finalizer.FinalizeResources(); err != nil {
+					fmt.Printf("error finalizing resources: %s", err.Error())
```
Should we be exiting with an error code here? (Or setting a flag so we exit with an error code when all the finalizers are done?)

Mr0grog commented on lib/badger.go:
```diff
@@ -0,0 +1,120 @@
+		Options: badger.DefaultOptions,
+	}
+	cfg.Dir = filepath.Join(wd, "badger")
+	cfg.ValueDir = filepath.Join(wd, "badger")
```
Hmmm, if there’s gonna be stateful stuff that sticks around here, we probably need to do something about how this should work/where it goes. For now, though, I feel like we should at least add this to `.gitignore`. (Or maybe when the process exits and cleans up, it should erase the Badger store?)

Mr0grog commented on lib/badger.go:
```diff
@@ -0,0 +1,120 @@
+// DB returns a connection to Badger
+func (c *BadgerConfig) DB() (conn *badger.DB, err error) {
+	if db == nil {
+		// Open the Badger database located in the /tmp/badger directory.
```
Oh, this doesn’t match the directory that is actually used. Was it supposed to be `/tmp/badger` above in `NewBadgerConfig()`? I feel like that would obviate my comment :)

Mr0grog commented on lib/config.go:
```diff
@@ -40,6 +54,15 @@ func DefaultConfig() *Config {
+// badger store if one doesn't exist
+func (cfg *Config) BadgerDB() (*badger.DB, error) {
+	if cfg.Badger == nil {
+		cfg.Badger = NewBadgerConfig()
```
Hmmm, I don’t feel super-great about this function modifying the config as a side-effect. I wonder if this would be better handled by setting this in the default config (and maybe utilizing the default config as a base when loading JSON configs).

Mr0grog commented on cmd/start.go:
```diff
@@ -36,9 +36,13 @@ var StartCmd = &cobra.Command{
-		// 	fmt.Printf("error writing file: %s", err.Error())
-		// }
+		for _, rh := range coord.ResourceHandlers() {
+			if finalizer, ok := rh.(lib.ResourceFinalizer); ok {
```
Hmmm, feels like this could be part of the coordinator interface, so you could either iterate over the finalizers:

```go
for _, finalizer := range coord.ResourceFinalizers() {
    if err := finalizer.FinalizeResources(); err != nil {
```

Or maybe just execute them:

```go
if errors := coord.FinalizeResources(); len(errors) != 0 {
```

Mr0grog replied:
(If you’re reading e-mail, I mistyped the first example here and just edited it on GitHub.)

Mr0grog commented on lib/resource.go:
```diff
@@ -110,7 +110,16 @@ func (u *Resource) HandleResponse(started time.Time, res *http.Response, recordH
+	if err != nil {
+		return "", err
+	}
+	return purell.NormalizeURL(u, purell.FlagsUnsafeGreedy), nil
```
Can this leverage `NormalizeURL(u)`?

Mr0grog commented on lib/resource_handler.go:
```diff
@@ -14,10 +15,17 @@ type ResourceHandler interface {
-func NewResourceHandlers(cfgs []*ResourceHandlerConfig) (rhs []ResourceHandler, err error) {
-	for _, c := range cfgs {
-		rh, err := NewResourceHandler(c)
+func NewResourceHandlers(cfg *Config) (rhs []ResourceHandler, err error) {
```
This doc comment is now inaccurate.

Mr0grog commented on lib/sitemap.go:
```diff
@@ -0,0 +1,203 @@
+	})
+
+	if err != nil {
+		log.Debug("error adding map entry to badger: %s", err.Error())
```
Should be `log.Debugf()`.

Mr0grog commented on lib/worker.go:
```diff
@@ -43,15 +42,18 @@ func NewWorker(cfg *WorkerConfig) (w Worker, err error) {
+	stop     chan bool
+	cfg      *WorkerConfig
+	fetchers []*fetchbot.Fetcher
+	queues   []*fetchbot.Queue
```
One the one hand, I feel like it would be better for the some other component to create multiple LocalWorkers instead of leaving it to the worker itself to manage its own parallelism, but on the other hand, I know the Chrome Headless one will benefit from working this way because it can manage a single browser process and have a separate tab for each parallel implementation.

So I guess this makes sense. I just wish it didn’t. :P

Mr0grog commented on lib/worker.go:
```diff
@@ -94,12 +100,18 @@ func (w *LocalWorker) Start(coord WorkCoordinator) error {
 				}
+				i++
+				if i == len(w.queues) {
+					i = 0
```
Could just be `i = (i + 1) % len(w.queues)`

b5 replied:
wow that's gorgeous

Mr0grog commented on lib/worker.go:
```diff
@@ -67,25 +69,29 @@ func (w *LocalWorker) SetDelay(d time.Duration) {
+	}
 
 	go func() {
+		i := 0
```
This isn’t different from the previous version, but it just occurred to me that this implementation will just automatically consume everything from the coordinator’s queue and put it in an internal queue here. Since the workers config is an array, don’t we need this to only consume from the coordinator’s queue when a fetchbot is free?

Mr0grog replied:
Thinking more about this:
- If one URL is especially slow to load, other URLs will get stuck behind it in the queue for the fetchbot handling it.
- Should the workers config even be an array? I'm trying to think of a situation where we’d *want* a heterogenous set of scraping algorithms (except where they *all* work every item in the queue, which is not how this component is set up to work).

b5 replied:
oh truth. This is interesting. I vote we work on this in another PR.

Mr0grog replied:
>> …Since the workers config is an array, don’t we need this to only consume from the coordinator’s queue when a fetchbot is free?
>
> …I vote we work on this in another PR.

👍

Mr0grog commented on lib/worker.go:
```diff
@@ -181,10 +193,10 @@ func NewRecordRedirectClient(wc WorkCoordinator) *http.Client {
 
 			r, _ := url.Parse(req.URL.String())
-			canurlstr := NormalizeURLString(r)
+			canurlstr := NormalizeURL(r)
```
Seems like you could just combine these two lines now?

```go
canurlstr, _ = NormalizeURLString(req.URL.String())
```

Mr0grog reviewed (changes requested):
Not a change from before, but just caught the fact that line 123 of `coordinator.go` should be `log.Errorf()` instead of `log.Error()`. (`go test` is screaming at me about it; I think it might be a change in v1.11.)

Otherwise, a bunch more minor notes inline. I’m not sure most of them are very serious, BUT I don want to come back around on my earlier point from the last review about needing to `ctrl+c` to stop walk here. It seems like the coordinator should definitely be able to determine when its queue has been successfully drained and worked through, so it should be able to know when it’s done and stop itself. That might be a bunch more work to do here, so if you want to push it off to another PR, can we at least note the `ctrl+c` requirement in the help text and log an issue for fixing it?

b5 committed:
refactor: incorporate PR #13 feedback

b5 committed:
fix(badger txn): update breaking badger txn value interface

b5 committed:
test(sitemap,cbor): add some tests like a self-respecting person

b5 committed:
chore(codecov,circleci): bump circle go image version, add codecov config

b5 committed:
fix(coordinator): fix logger error call

b5 commented:
a'ight @Mr0grog, you up

Mr0grog commented on lib/resource_handler_test.go:
```diff
@@ -0,0 +1,34 @@
+	rh.HandleResource(r)
+
+	r.Hash = "not_actually_a_hash"
+	rh.HandleResource(r)
```
Should these two calls to `rh.HandleResource()` be followed by an `os.Stat()` to verify that a file was [not] written?

Mr0grog commented on lib/badger.go:
```diff
@@ -0,0 +1,119 @@
+	// wd, err := os.Getwd()
+	// if err != nil {
+	// 	panic(err)
+	// }
```
I’m assuming that if the values for `Dir` and `ValueDir` don’t start with `/`, they are already interpreted as relative to the working dir? ([Docs are not clear](https://godoc.org/github.com/dgraph-io/badger#Options).) Why are these commented out and still here?

(Same Q for the code that uses this, a couple lines down.)

Mr0grog commented on lib/config.go:
```diff
@@ -36,10 +50,24 @@ func DefaultConfig() *Config {
+// badger store if one doesn't exist
+func (cfg *Config) BadgerDB() (*badger.DB, error) {
+	if cfg.Badger == nil {
+		return nil, ErrNoBadgerConfig
```
In a previous version of this code, we *created* a new Badger config here. I think this change is mostly reasonable except it exposes an issue I hadn’t noticed before: `JSONConfigFromFilepath` completely overwrites instead of adding to the existing config. https://github.com/qri-io/walk/blob/e13838a20af9a993969e689c7d581846df265f18/lib/config.go#L43-L59

Because of that, it took me a bit of hunting to realize I now need to add  a `Badger` section to my config file… and that that section needs more than just the `Dir` and `ValueDir` properties. It doesn’t seem to match up with what the [Badger docs say is required](https://godoc.org/github.com/dgraph-io/badger#Options), so a after a couple false starts, I just copied the entirety of `badger.DefaultOptions` in there:

```js
{
    "Coordinator": /* ... */,
    "Queue": /* ... */,
    "RequestStore": /* ... */,
    "Workers": /* ... */,
    "ResourceHandlers": /* ... */,
    "Badger": {
      "Dir": "badger",
      "ValueDir": "badger",
      "ValueLogFileSize": 1073741823,
      "DoNotCompact":        false,
      "LevelOneSize":        268435456,
      "LevelSizeMultiplier": 10,
      "TableLoadingMode":    1,  // options.LoadToRAM, not so great to reference an enum/constant this way
      "ValueLogLoadingMode": 2,  // options.MemoryMap, not so great to reference an enum/constant this way

      "MaxLevels":               7,
      "MaxTableSize":            67108864,
      "NumCompactors":           3,
      "NumLevelZeroTables":      5,
      "NumLevelZeroTablesStall": 10,
      "NumMemtables":            5,
      "SyncWrites":              true,
      "NumVersionsToKeep":       1,

      "ValueLogMaxEntries": 1000000,
      "ValueThreshold":     32,
      "Truncate":           false
    }
}
```

Anyway, all that seems pretty painful. (Also: was there an easier way for me to have done this?) I think it’s probably `JSONConfigFromFilepath` that should change and the code you have here now should stay, but not really certain about that.

Barring any other changes, it seems like we should at least document this in an example config file.

Mr0grog reviewed (approved):
This all works and seems fine to me for the most part. I do have big concerns about config (noted inline below). I’d *like* to address that first (if not by code changes that at least by adding a `config.example.json` since it’s gotten so complex), but am not going to consider that blocking since this has been sitting for so long.

A few other comments inline, too.

b5 committed:
feat(config commnad): add config command to see what config walk is using

b5 commented:
This isn't a direct-fix for your config worries, but I think a more sustainable first step toward demystifing config: I've added a new command `walk config` that prints out the current config walk is using, which will be the default if no config is present.

I think we should at least add documentation about how to use this command to understand configuration, but is this sufficient to merge this branch @Mr0grog?

Mr0grog commented:
Hmmm, I don’t think this addresses my concerns about configuring *Badger*, though, which is the complicated part (because those config options have to be looked up from outside this codebase).

Mr0grog commented:
(If you do `./walk config` it reads in the `config.json` you added, which doesn’t configure Badger, and so it doesn’t output any Badger config. That’s all fine if you aren’t using the `sitemap` resource handler, but as soon as you do, this command won’t really help you figure out what to add unless you know you have to first totally remove all config files.)

b5 commented:
Ah yes, ok, mind if we merge this & open up an issue to discuss. I totally agree with your points, and think we should slow down & write a proper fix instead of tacking stuff onto a month-old PR 😉

Mr0grog commented:
👍 works for me.

b5 commented:
lovely
